packages/next/server/next-server.ts

```diff
@@ -1333,3 +1352,74 @@ export default class NextNodeServer extends BaseServer {
     }
   }
 }
+
+/**
+ * Creates a ReadableStream from a Node.js HTTP request
+ */
+function requestToBodyStream(request: IncomingMessage): BodyStream {
+  const transform = new TransformStream<Uint8Array, Uint8Array>({
+    start(controller) {
+      request.on('data', (chunk) => controller.enqueue(chunk))
+      request.on('end', () => controller.terminate())
+      request.on('error', (err) => controller.error(err))
+    },
+  })
+
+  return transform.readable as unknown as ReadableStream<Uint8Array>
+}
+
+/**
+ * A simple utility to take an original stream and have
+ * an API to duplicate it without closing it or mutate any variables
+ */
+function teeableStream<T>(originalStream: ReadableStream<T>): {
+  duplicate(): ReadableStream<T>
+  original(): ReadableStream<T>
+} {
+  return {
+    duplicate() {
+      const [stream1, stream2] = originalStream.tee()
+      originalStream = stream1
+      return stream2
+    },
+    original() {
+      return originalStream
+    },
+  }
+}
+
+function bodyStreamToNodeStream(bodyStream: BodyStream): NodeStreams.Readable {
+  const reader = bodyStream.getReader()
+  return NodeStreams.Readable.from(
+    (async function* () {
+      while (true) {
+        const { done, value } = await reader.read()
+        if (done) {
+          return
+        }
+        yield value
+      }
+    })()
+  )
+}
+
+function enhanceIncomingMessage<T extends IncomingMessage>(
+  base: T,
+  body: BodyStream
+): T {
+  const stream = bodyStreamToNodeStream(body)
+  return new Proxy<T>(base, {
+    get(target, name) {
+      if (name in stream) {
+        const v = stream[name as keyof NodeStreams.Readable]
+        if (typeof v === 'function') {
+          return v.bind(stream)
+        } else {
+          return v
+        }
+      }
+
+      return target[name as keyof T]
+    },
+  })
+}
```

Comment:
I know that it is super weird. I need to do something to replace the original body (that was read already by the middlewares). We can obviously make it so we only replace if it was actually read as an optimization but we have to create a new request (Node.js' `IncomingMessage`) to allow reading "new data".

There's another way of doing that: by mutating the original request.
We then can call this function `replaceRequestBody`:

```tsx
function replaceRequestBody<T extends IncomingMessage>(
  base: T,
  body: BodyStream
): T {
  const stream = bodyStreamToNodeStream(body)

  for (const key in stream) {
    let v = stream[key as keyof NodeStreams.Readable] as any
    if (typeof v === 'function') {
      v = v.bind(stream)
    }
    base[key as keyof T] = v
  }

  return base
}
```

Maybe mutating makes more sense as it's more concise, more "JavaScript-y" and avoids Proxies?